.test-infra/jenkins/Flink.groovy

```diff
@@ -19,10 +19,74 @@
 import CommonJobProperties as common
 import CommonTestProperties.SDK
 
-class Infrastructure {
+class Flink {
+  private static final String repositoryRoot = 'gcr.io/apache-beam-testing/beam_portability'
+  private static final String dockerTag = 'latest'
+  private static final String jobServerImageTag = "${repositoryRoot}/flink-job-server:${dockerTag}"
+  private static final String flinkVersion = '1.7'
+  private static final String flinkDownloadUrl = 'https://archive.apache.org/dist/flink/flink-1.7.0/flink-1.7.0-bin-hadoop28-scala_2.11.tgz'
+
+  private static def job
+  private static String jobName
+
+ /**
+  * Returns SDK Harness image tag to be used as an environment_config in the job definition.
+  *
+  * @param sdk - SDK
+  */
+  static String getSDKHarnessImageTag(SDK sdk) {
+    switch (sdk) {
+      case CommonTestProperties.SDK.PYTHON:
+        return "${repositoryRoot}/python:${dockerTag}"
+      case CommonTestProperties.SDK.JAVA:
+        return "${repositoryRoot}/java:${dockerTag}"
+      default:
+        String sdkName = sdk.name().toLowerCase()
+        throw new IllegalArgumentException("${sdkName} SDK is not supported")
+    }
+  }
+
+  /**
+   * Creates Flink cluster and specifies cleanup steps.
+   *
+   * @param job - jenkins job
+   * @param jobName - string to be used as a base for cluster name
+   * @param sdk - SDK
+   * @param workerCount - the initial number of worker nodes excluding one extra node for Flink's Job Manager
+   * @param slotsPerTaskmanager - the number of slots per Flink task manager
+   */
+  static Flink setUp(job, String jobName, SDK sdk, Integer workerCount, Integer slotsPerTaskmanager = 1) {
```

Comment:
Maybe it's even better to create a separate class for supplying docker images, e.g. Portability.groovy? I think this should be reconsidered as well. In the future, any other portable cluster will need these images. Maybe it's good to extract the responsibility of building & pushing them? WDYT?

This is exciting! :) If we cut this nicely it will be super easy to use!

Reply:
> Maybe it's even better to create a separate class for supplying docker images, e.g. Portability.groovy?
That was an interesting suggestion, thank you. I decided to extract those responsibilities and create new classes `DockerPublisher` and `SDKHarnessPublisher`.  DockerPublisher has a public `publish` method which takes gradle task and the name and the tag of an image as an arguments. `SDKHarnessPublisher` has a similar method which takes only the SDK.

> sometimes users of this interface may need a cluster without sdk harness and the job server

Fair point. I reorganised Flink's interface; now it has the following methods:
- public constructor for initializing the object
- prepareJobServer
- setUp - specifies start and teardown steps